### PR TITLE
ConfigureEnvironment big refactor

### PR DESCRIPTION
**Pull request by lasote**
- Merge first: #989
- ConfigureBuildEnvironment with environment_append
- New virtualbuildenv for build variables. #408 
- Appveyor: Added mingw64 cached (to avoid to much traffic for s3).
- Travis: Increased logger level, osx builds failing because of max log size (temporary fix?)
- Travis: Installed some x86 libs for building with gcc x86 apps
- Added some traces to output when user conanfile build() method fails.

### CODE REVIEW

File: conans/tools.py [hunk 1]
```diff
@@ -1,46 +1,104 @@
...
+    except:
+        arch = None
+    env_vars = {"MSYSTEM": "MINGW32" if arch == "x86" else "MINGW64",
+                "MSYS2_PATH_TYPE": "inherit"}
+    with environment_append(env_vars):
+        curdir = unix_path(cwd or os.path.abspath(os.path.curdir))
+        # Needed to change to that dir inside the bash shell
+        to_run = 'cd "%s" && %s ' % (curdir, bashcmd)
+        wincmd = 'bash --login -c %s' % escape_windows_cmd(to_run)
+        conanfile.output.warn('run_msys2_cmd: %s' % wincmd)
```
Comment (16k) [inline conans/tools.py]: output should probably be "run_in_windows_bash" instead of "run_msys2_cmd".

Comment (lasote) [inline conans/tools.py]: Also will change from warn to info. Thanks

File: conans/client/generators/gcc.py [hunk 2]
```diff
@@ -12,17 +15,39 @@ def content(self):
...
+        rpath_separator = "," if platform.system() == "Darwin" else "="
+        flags.extend(['-Wl,-rpath%s"%s"' % (rpath_separator, x.replace("\\", "/")) 
+                      for x in self._deps_build_info.lib_paths]) # rpaths
+        flags.extend(['-L"%s"' % x.replace("\\", "/") for x in self._deps_build_info.lib_paths])
+        flags.extend(["-l%s" % x for x in self._deps_build_info.libs])
+        flags.extend(self._deps_build_info.cppflags)
+        flags.extend(self._deps_build_info.cflags)
+        flags.extend(self._deps_build_info.sharedlinkflags)
+        flags.extend(self._deps_build_info.exelinkflags)
+        flags.extend(self._libcxx_flags())
```
Comment (memsharded) [inline conans/client/generators/gcc.py]: Seems some duplication with AutoToolsEnvironment, in _libcxx_flags and architecture, please check if makes sense to refactor.

File: conans/client/generators/gcc.py [hunk 3]
```diff
@@ -12,17 +15,39 @@ def content(self):
...
+
+        lib_flags = []
+        if libcxx:
+            if libcxx == "libstdc++":
+                lib_flags.append("-D_GLIBCXX_USE_CXX11_ABI=0")
+            elif str(libcxx) == "libstdc++11":
+                lib_flags.append("-D_GLIBCXX_USE_CXX11_ABI=1")
+
+            if "clang" in compiler:
+                lib_flags.append({"libc++": "-stdlib=libc++"}.get(libcxx, "-stdlib=libstdc++"))
```
Comment (memsharded) [inline conans/client/generators/gcc.py]: Shouldn't this libc++ case be also handled in AutoToolsEnvironment?

File: conans/test/integration/build_environment_test.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    #ifdef ACTIVE_VAR
+    std::cout << "Active var!!!" << std::endl;
+    #endif
+}
+    '''
+
+
+class BuildEnvironmenTest(unittest.TestCase):
+
+    def test_autotools_environment(self):
```
Comment (memsharded) [inline conans/test/integration/build_environment_test.py]: Remove or TODO?

File: conans/client/generators/gcc.py (hunk 3 above)
Comment (lasote) [inline conans/client/generators/gcc.py]: It is, but now I append definitions with definitions and flags with flags, not all together.

File: conans/test/integration/build_environment_test.py (hunk 4 above)
Comment (lasote) [inline conans/test/integration/build_environment_test.py]: remove, already tested with basic_build_test, diamond and some other with the cpp_test_files

**Pull request merged**